### PR TITLE
[DC-1329] create sandbox utility options function

### PR DESCRIPTION
**Pull request by jp3477**
Creates an options clause for defining BigQuery table descriptions and labels.

### CODE REVIEW

File: data_steward/utils/sandbox.py [hunk 1]
```diff
@@ -63,3 +89,76 @@ def check_and_create_sandbox_dataset(project_id, dataset_id):
...
+    return TABLE_LABELS_STRING.render(labels=labels)
+
+
+def get_sandbox_table_description_string(description):
+    """A helper function that returns a formatted description for BigQuery
+
+    :param str description: A table description
+    :return: A formatted table description
+    :rtype: str
+    """
```
Comment (amcgrenera-vumc) [inline data_steward/utils/sandbox.py:140]: Do you know if this is required? Not sure if there should be a check in this function that something was actually passed?

File: tests/unit_tests/data_steward/utils/sandbox_test.py [hunk 2]
```diff
@@ -29,3 +29,42 @@ def test_get_sandbox_table_name(self):
...
+                dataset_name,
+                empty_class_name,
+                table_tag,
+                shared_lookup=non_bool_shared_lookup)
+
+    def test_get_sandbox_table_description_string(self):
+        description = "This is a helpful description."
+
+        actual = sandbox.get_sandbox_table_description_string(description)
+        self.assertIn(f'description="{description}"', actual)
```
Comment (amcgrenera-vumc) [inline tests/unit_tests/data_steward/utils/sandbox_test.py:70]: Should there be one more unit test to test the `get_sandbox_options` to ensure the `TABLE_OPTIONS_CLAUSE` is filled out correctly?

File: data_steward/utils/sandbox.py [hunk 3]
```diff
@@ -63,3 +89,76 @@ def check_and_create_sandbox_dataset(project_id, dataset_id):
...
+                        class_name,
+                        table_tag,
+                        desc,
+                        shared_lookup=False):
+    """A function that assembles a BigQuery table options clause from labels and descriptions
+
+    :param str dataset_name: A dataset name
+    :param str class_name: A class name
+    :param str table_tag: A table tag
+    :param str description: A table description
```
Comment (amcgrenera-vumc) [inline data_steward/utils/sandbox.py:154]: nit: change to `desc` to match arg

File: data_steward/utils/sandbox.py [hunk 4]
```diff
@@ -63,3 +89,76 @@ def check_and_create_sandbox_dataset(project_id, dataset_id):
...
+    :param str class_name: A class name
+    :param str table_tag: A table tag
+    :param bool shared_lookup: A boolean that indicates if a table is a shared lookup, defaults to False
+    :raises ValueError: Raises if one of first three arguments is whitespace or None
+    :raises ValueError: Raises if the `shared_lookup` is not a boolean
+    :return: A formatted string of the arguments concatenated as labels
+    :rtype: str
+    """
+    STRING_LABELS = ['src_dataset', 'class_name', 'table_tag']
+    BOOL_LABELS = ['shared_lookup']
```
Comment (amcgrenera-vumc) [inline data_steward/utils/sandbox.py:110]: nit: maybe make this lowercase since they are instance variables

Comment (jp3477) [reply data_steward/utils/sandbox.py:110]: Sounds good

File: data_steward/utils/sandbox.py (hunk 1 above)
Comment (jp3477) [reply data_steward/utils/sandbox.py:140]: I'm not sure. In the task, it doesn't specifically mention a required check for description. But I can add it if you think that's best.

File: data_steward/utils/sandbox.py (hunk 3 above)
Comment (jp3477) [reply data_steward/utils/sandbox.py:154]: Good catch

File: tests/unit_tests/data_steward/utils/sandbox_test.py (hunk 2 above)
Comment (jp3477) [reply tests/unit_tests/data_steward/utils/sandbox_test.py:70]: I agree, wasn't sure how I would go about this, but I have an idea now.

File: data_steward/utils/sandbox.py [hunk 5]
```diff
@@ -63,3 +89,82 @@ def check_and_create_sandbox_dataset(project_id, dataset_id):
...
+    """A helper function that returns a formatted description for BigQuery
+
+    :param str description: A table description
+    :return: A formatted table description
+    :rtype: str
+    """
+
+    if not description or description.isspace():
+        raise ValueError(
+            f"Description must be a non-empty string. Received teh value '{description}'."
```
Comment (amcgrenera-vumc) [inline data_steward/utils/sandbox.py:144]: nit: spell check `the`

Comment (jp3477) [reply data_steward/utils/sandbox.py:144]: Oops, fixing

File: tests/unit_tests/data_steward/utils/sandbox_test.py [hunk 6]
```diff
@@ -29,3 +30,88 @@ def test_get_sandbox_table_name(self):
...
+
+        #Test description empty
+        description = "    "
+        with self.assertRaises(ValueError):
+            sandbox.get_sandbox_table_description_string(description)
+
+    @patch('utils.sandbox.get_sandbox_labels_string')
+    @patch('utils.sandbox.get_sandbox_table_description_string')
+    def test_get_sandbox_options(self,
```
Comment (amcgrenera-vumc) [inline tests/unit_tests/data_steward/utils/sandbox_test.py:80]: this looks great, thanks for adding in!

**Comment by amcgrenera-vumc** [comment]
This all looks great. One last thing! I was testing out the option string in BQ to make sure it would run properly and I found that the values for the labels need to be all lower case: https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#labels

> The labels associated with this table. You can use these to organize and group your tables. Label keys and values can be no longer than 63 characters, **can only contain lowercase letters**, numeric characters, underscores and dashes. International characters are allowed. Label values are optional. Label keys must start with a letter and each label in the list must have a different key.

So for the `shared_lookup`, I think we can still keep it a boolean but when adding the value convert to string and  `.lower()`? let me know your thoughts!

**Comment by jp3477** [comment]
> This all looks great. One last thing! I was testing out the option string in BQ to make sure it would run properly and I found that the values for the labels need to be all lower case: https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#labels
> 
> > The labels associated with this table. You can use these to organize and group your tables. Label keys and values can be no longer than 63 characters, **can only contain lowercase letters**, numeric characters, underscores and dashes. International characters are allowed. Label values are optional. Label keys must start with a letter and each label in the list must have a different key.
> 
> So for the `shared_lookup`, I think we can still keep it a boolean but when adding the value convert to string and `.lower()`? let me know your thoughts!

Nice catch, I didn't realize this at all. Sure, that sounds like the best plan.

**Review by amcgrenera-vumc — APPROVED** [review]
Looks great, thanks for making the changes! Once the integration test passes good to merge.

**Pull request merged**